rfcs/012-fdb-reduce.md

```diff
@@ -0,0 +1,190 @@
+
+![figure 3: Traversing flow to return results for `group_level = 2`](images/SkExample3.png)
+*Figure 3:* Traversing flow to return results for `group_level = 2`
+
+A query with startkey/endkey would follow a similar process.  Start at the highest level traversing across until we exceed the startkey, then move down until we find the startkey or nearest node to it. Then follow the above process of traversing the skip list to return all results until we reach a node that is greater or equal to the endkey.
+
+#### Update
+
+To update the reduce index, we will use the same map id index that keeps track of what keys are associated with a document. When a document is deleted. The previous keys will be removed from level 0. If the reduce function is `_sum` or `_count`, an atomic update is then performed for all Node's above Level 0 that would have included the values for the keys deleted.  
+
+For reduce functions where we cannot perform an atomic update. The process for each level, above level 0, is to fetch all the key/values in the level below the current node that are used to calculate the current node's aggregation value. Re-reduce those keys to calculate the new value stored for the node and store those results back in FoundationDB.
+
+When updating a document, the initial delete process is followed to remove the existing keys that are no longer emitted for this document. The new keys are added at level 0. For level's above 0, the same distribution algorithm will be used to determine if the key/values are added to a level. If they are, then an aggregation of the nodes after this node but at the level below is performed to calculate the aggregation value stored for this node. The previous node's value is also recalculated. This continues up to the max level on the skip list. For `_sum` and `_count` this can be done as an atomic update. For other reduce functions this will be a re-reduce. If the new key/value is not added to the level, the value is aggregated with a node smaller than the key in that level.
+
+### Data model
+
+The data model for the skip list implementation will follow the same design as map indexes where we store one row for the key and one row for the Value.
```

Comment:
I went back and read the map index RFC but didn't see any rationale for why we do this. It seems pretty wasteful. Is it because you want to reserve the full 100 KB for a possible emitted value?

Reply:
Good point. I need to update the map index to reflect this information. The reason we do this is once we encode the key so that the index sorts correctly. We cannot always decode it to exactly way the key that was emitted. Specifically for numbers which are encoded to doubles when stored in fdb but when returned to the user should often be integers. You had the discussion around number conversions on the mailing list. If we could convert all numbers to doubles I think we could remove the extra row.

Reply:
Sorry, I should have been more specific. I know why we have to store the originally-emitted Key into an FDB Value. What I don't understand is why we dedicated a whole Value just to storing the emitted Key instead of packing it in with the emitted Value into a single FDB Value?

Reply:
Ok right. Good point. Given that reduce values should be small. It does make a lot of sense to add the key and value in the same row. Thanks. I'll update the RFC.

Reply:
Cool. For what it's worth the design you propose may not have the same "death spiral" effect with large reductions that CouchDB has today. The reason that happens today is because the btree chooses when to split based on the size of the inner node. Large reductions lead to a reduction of the branching factor and a deepening of the tree, which also causes more copies of the large reduction to be stored. And down we go :)

If you're careful in how you choose skip list levels we may end up much less sensitive to that effect.